examples/Scikit.ipynb

```diff
@@ -0,0 +1,330 @@
+      "text/plain": [
+       "<Figure size 720x360 with 10 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "import time\n",
+    "from sklearn.model_selection import train_test_split\n",
+    "from sklearn.linear_model import LogisticRegression\n",
+    "\n",
+    "t0 = time.time()\n",
+    "\n",
+    "# Split data into test and train sets\n",
+    "train_samples = 20000\n",
+    "\n",
+    "X_train, X_test, y_train, y_test = train_test_split(\n",
+    "    X, y, train_size=train_samples, test_size=10000)\n",
+    "\n",
+    "# Define model, lower C value gives higher regulation\n",
+    "clf = LogisticRegression(\n",
+    "    C=0.001, penalty='l2', solver='saga', tol=0.1\n",
+    ")\n",
+    "\n",
+    "# Fit model\n",
+    "clf.fit(X_train, y_train)\n",
+    "\n",
+    "# Score model\n",
+    "score = clf.score(X_test, y_test)\n",
+    "print(\"Test score with L2 penalty: %.4f\" % score)\n",
+    "\n",
+    "# Visualize coefficients\n",
+    "coef = clf.coef_.copy()\n",
+    "max_val = np.abs(coef).max()\n",
+    "\n",
+    "plt.figure(figsize=(10, 5))\n",
+    "for i in range(10):\n",
+    "    p = plt.subplot(2, 5, i + 1)\n",
+    "    p.imshow(coef[i].reshape(28, 28), cmap=plt.cm.RdBu,\n",
+    "             vmin=-max_val, vmax=max_val)\n",
+    "    p.set_xticks(())\n",
+    "    p.set_yticks(())\n",
+    "    p.set_xlabel('Class %i' % i)\n",
+    "plt.suptitle('Classification vector for...')\n",
+    "\n",
+    "run_time = time.time() - t0\n",
+    "print('Example run in %.3f s' % run_time)\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "52a034e3",
+   "metadata": {},
+   "source": [
+    "## The \"Application\"\n",
+    "\n",
+    "We can use the `xaitk-saliency` high-level API to generate saliency maps that compare to the visualization above.\n",
+    "\n",
+    "Our \"application\" will accept a set of images, a blackbox image classifier, and a saliency generator and will generate saliency maps for each image provided.\n",
+    "It will then combine the heatmaps of each corresponding class and plot them.\n",
+    "This should give us a view of the decision boundaries for each class similar to the one above."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "08fb1a3e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from smqtk_classifier import ClassifyImage\n",
+    "from xaitk_saliency import GenerateImageClassifierBlackboxSaliency\n",
+    "from functools import reduce\n",
+    "from math import ceil\n",
+    "\n",
+    "def app(\n",
+    "    images: np.ndarray,\n",
+    "    image_classifier: ClassifyImage,\n",
+    "    saliency_generator: GenerateImageClassifierBlackboxSaliency\n",
+    "):\n",
+    "    num_imgs = images.shape[0]\n",
+    "    \n",
+    "    # Generate saliency maps\n",
+    "    sal_maps = []\n",
+    "    for i in range(num_imgs):    \n",
```

Comment:
I think here, it would be nice to also visualize saliency maps for individual MNIST examples. We could make the story to be: 1) as a form of "local" explanation, show saliency maps for example digit exemplars, and 2) as a form of "global" explanation, show average saliency maps for each digit class.

Related to this, it is not surprising to me that the saliency maps agree with the coefficients learned by the logistic regression model since we are operating in the linear regime (i.e. each coefficient is a measure of influence on the model's final prediction, and since each coefficient maps to a certain pixel, this by definition maps directly to a pixel's influence on the final model prediction).

One thing that could be potentially interesting to try would be to borrow the MLP from the second scikit-learn [example](https://scikit-learn.org/stable/auto_examples/neural_networks/plot_mnist_filters.html#sphx-glr-auto-examples-neural-networks-plot-mnist-filters-py) (which is more than 1 layer) and apply our saliency maps there. The model is more complex and also achieves higher accuracy. I believe in that example they visualize the weights of the first layer, which are not great. This would be a slightly more involved example, but I still think instructive for the reader. @Purg Any thoughts on this?

Reply:
Adding a follow-on MLP take sounds like a good idea to me. In adding the MLP example this into *this* notebook, I don't think we need to reproduce the reference example, just refer to it (since their results are there). We can simply create the `MLPClassifier` instance, fit it, and apply the model as a blackbox to our "app" for saliency visualization.

Side note, even their example states that the MLP model will warn that it does not converge, so it doesn't surprise me that their visualization might not be that great (it's also just the first layer out of 50, so I'm not sure what they were expecting to get out of that visualization...). We can either duplicate their hyperparameters to provide a comparable result, or tweak the iteration/learning-rate to "better" the model for our visualization. I might suggest the former for comparability while noting that we understand that non-convergence is happening.